File: Codeforces/Practice/1059D.py
import sys
import math
from decimal import Decimal

n = int(input())
point = []
neg = False
pos = False
for i in range(n):
	x, y = sys.stdin.readline().split()
	x = int(x)
	y = int(y)
	if y < 0:
		neg = True
		y = y * -1
	else:
		pos = True
	point.append((x, y))
if neg is True and pos is True:
	print('-1')
	sys.exit(0)
lo = Decimal(0)
hi = Decimal(1e16)
ans = True
ctr = 0
while hi - lo >= 1e-8 and ctr < 200:
	mid = Decimal(lo + hi)/Decimal(2.0)
	#mid = Decimal.from_float(mid)
	#print(mid)
	x1 = -1e16
	x2 = 1e16
	iter_ans = True
	# print(ctr, mid, lo, hi)
	for i in range(n):
		vvv = Decimal(Decimal(2 * mid) - Decimal(point[i][1]))
		vvv = Decimal(vvv * Decimal(point[i][1]))
		if(vvv < 0):
			iter_ans = False
			break
		v2 = Decimal(vvv).sqrt()
		v1 = Decimal(v2 * -1)
		v1 = Decimal(v1 + point[i][0])
		v2 = Decimal(v2 + point[i][0])
		if v2 < x1 or v1 > x2:
			iter_ans = False
			break
		x1 = max(x1, v1)
		x2 = min(x2, v2)
	# print(iter_ans)
	if iter_ans == True:
		hi = mid
	else:
		lo = mid + Decimal(1e-8)
	ans = iter_ans
	ctr += 1

if lo != 1e16:
	print(Decimal(lo))
else:
	print('-1')

File: Codeforces/Practice/1059D.cpp
#include <bits/stdc++.h>
using namespace std;

struct node	{
	int x, y;
}point[100100];

int main()	{
	int n;
	cin >> n;
	bool pos = false;
	bool neg = false;
	for(int i = 0; i < n; i++)	{
		cin >> point[i].x >> point[i].y;
		if(point[i].y > 0)	{
			pos = true;
		}
		else	{
			neg = true;
			point[i].y *= -1;
		}
	}
	if(pos && neg)	{
		cout << "-1\n";
		return 0;
	}
	double lo = 0;
	double hi = 1e15;
	double mid;
	double x1, x2;
	int ctr = 0;
	while(hi - lo >= 1e-8 && ctr < 200)	{
		mid = (lo + hi)/2.0;
		x1 = -1e16, x2 = 1e16;
		bool temp_ans = true;
		//cout << setprecision(13) << std::fixed << mid << endl;
		for(int i = 0; i < n; i++)	{
			double chh = 2 * mid - point[i].y;
			chh *= point[i].y;
			if(chh < 0)	{
				temp_ans = false;
				break;
			}
			double v2 = (double)sqrt(chh);
			double v1 = v2 * -1;
			v1 += point[i].x;
			v2 += point[i].x;
			if(v1 > x2 || v2 < x1)	{
				temp_ans = false;
				break;
			}
			x1 = max(x1, v1);
			x2 = min(x2, v2);
		}
		if(temp_ans)	{
			hi = mid;
			//cout << "True\n";
		}
		else	{
			lo = mid + 1e-8;
			//cout << "False\n";
		}
		ctr++;
	}
	if(lo != 1e16)	{
		cout << setprecision(13) << std::fixed << lo << endl;
	}
	else	{
		cout << "-1" << endl;
	}
	return 0;
}